### PR TITLE
Implement support for real Xbox keys

### PR DESCRIPTION
**Pull request by LukeUsher**
This allows for Cxbx-Reloaded to use the real certificate and eeprom keys from a real xbox.
This opens up the possibility of sharing savedata between the two systems.

For legal reasons, we cannot include these keys within Cxbx-Reloaded, instead they need to be dumped from a real xbox using the following tool, and copying the generated keys.bin file to the AppData folder.

https://github.com/JayFoxRox/xqemu-tools/blob/a342fa60234f6c9d01e216591435195946a5784f/dump-xbox.c

### CODE REVIEW

File: import/OpenXDK/include/xboxkrnl/xbox.h [hunk 1]
```diff
@@ -424,11 +428,11 @@ XBSYSAPI EXPORTNUM(353) XBOX_KEY_DATA XboxLANKey;
 // ******************************************************************
 // * 0x0162 - XboxAlternateSignatureKeys
 // ******************************************************************
-XBSYSAPI EXPORTNUM(354) XBOX_KEY_DATA XboxAlternateSignatureKeys;
+XBSYSAPI EXPORTNUM(354) XBOX_KEY_DATA XboxAlternateSignatureKeys[16];
 
 // ******************************************************************
 // * 0x0163 - XePublicKeyData
-// ******************************************************************
+// ****************1**************************************************
```
Comment (PatrickvL) [inline import/OpenXDK/include/xboxkrnl/xbox.h]: 1

File: import/OpenXDK/include/xboxkrnl/xbox.h [hunk 2]
```diff
@@ -424,11 +428,11 @@ XBSYSAPI EXPORTNUM(353) XBOX_KEY_DATA XboxLANKey;
 // ******************************************************************
 // * 0x0162 - XboxAlternateSignatureKeys
 // ******************************************************************
-XBSYSAPI EXPORTNUM(354) XBOX_KEY_DATA XboxAlternateSignatureKeys;
+XBSYSAPI EXPORTNUM(354) XBOX_KEY_DATA XboxAlternateSignatureKeys[16];
```
Comment (PatrickvL) [inline import/OpenXDK/include/xboxkrnl/xbox.h]: constantify 16

File: src/CxbxKrnl/CxbxKrnl.cpp [hunk 3]
```diff
@@ -163,6 +163,28 @@ void ApplyMediaPatches()
...
+	// Set the LAN Key
+	xboxkrnl::XcHMAC(xboxkrnl::XboxCertificateKey, xboxkrnl::XBOX_KEY_LENGTH, pCertificate->bzLanKey, xboxkrnl::XBOX_KEY_LENGTH, NULL, 0, Digest);
+	memcpy(xboxkrnl::XboxLANKey, Digest, xboxkrnl::XBOX_KEY_LENGTH);
+
+	// Signature Key
+	xboxkrnl::XcHMAC(xboxkrnl::XboxCertificateKey, xboxkrnl::XBOX_KEY_LENGTH, pCertificate->bzSignatureKey, xboxkrnl::XBOX_KEY_LENGTH, NULL, 0, Digest);
+	memcpy(xboxkrnl::XboxSignatureKey, Digest, xboxkrnl::XBOX_KEY_LENGTH);
+
+	// Alternate Signature Keys
+	for (int i = 0; i < 16; i++) {
```
Comment (PatrickvL) [inline src/CxbxKrnl/CxbxKrnl.cpp]: constantify 16

File: src/CxbxKrnl/CxbxKrnl.cpp [hunk 4]
```diff
@@ -477,6 +499,38 @@ void CxbxKrnlMain(int argc, char* argv[])
...
+		long size = ftell(fp);
+		rewind(fp);
+
+		// If the size of Keys.bin is correct (two keys), read it
+		if (size == xboxkrnl::XBOX_KEY_LENGTH * 2) {
+			fread(keys, xboxkrnl::XBOX_KEY_LENGTH, 2, fp);
+
+			memcpy(xboxkrnl::XboxEEPROMKey, &keys[0], xboxkrnl::XBOX_KEY_LENGTH);
+			memcpy(xboxkrnl::XboxCertificateKey, &keys[1], xboxkrnl::XBOX_KEY_LENGTH);
+		}
```
Comment (PatrickvL) [inline src/CxbxKrnl/CxbxKrnl.cpp]: else log

File: src/CxbxKrnl/EmuKrnlXbox.cpp [hunk 5]
```diff
@@ -81,23 +79,23 @@ XBSYSAPI EXPORTNUM(324) xboxkrnl::XBOX_KRNL_VERSION xboxkrnl::XboxKrnlVersion =
 
 // ******************************************************************
 // * 0x0145 - XboxSignatureKey
+// Generated programatically using the Certificate Key and the XBE's
```
Comment (PatrickvL) [inline src/CxbxKrnl/EmuKrnlXbox.cpp]: instead of programmatically, mention actual function doing the work

File: src/CxbxKrnl/EmuKrnlXbox.cpp [hunk 6]
```diff
@@ -81,23 +79,23 @@ XBSYSAPI EXPORTNUM(324) xboxkrnl::XBOX_KRNL_VERSION xboxkrnl::XboxKrnlVersion =
...
-{
-	// cxbx default saved game key
-	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
-};
+XBSYSAPI EXPORTNUM(325) xboxkrnl::XBOX_KEY_DATA xboxkrnl::XboxSignatureKey = { 0 };
 
 // ******************************************************************
 // * 0x0161 - XboxLANKey
+// Generated programatically using the Certificate Key and the XBE's
```
Comment (PatrickvL) [inline src/CxbxKrnl/EmuKrnlXbox.cpp]: same

File: src/CxbxKrnl/EmuKrnlXbox.cpp [hunk 7]
```diff
@@ -81,23 +79,23 @@ XBSYSAPI EXPORTNUM(324) xboxkrnl::XBOX_KRNL_VERSION xboxkrnl::XboxKrnlVersion =
...
+// Generated programatically using the Certificate Key and the XBE's 
+// LAN Key
 // ******************************************************************
-// TODO : What should we initialize this to?
-XBSYSAPI EXPORTNUM(353) xboxkrnl::XBOX_KEY_DATA xboxkrnl::XboxLANKey = {};
+XBSYSAPI EXPORTNUM(353) xboxkrnl::XBOX_KEY_DATA xboxkrnl::XboxLANKey = { 0 };
 
 // ******************************************************************
 // * 0x0162 - XboxAlternateSignatureKeys
+// Generated programatically using the Certificate Key and the XBE's
```
Comment (PatrickvL) [inline src/CxbxKrnl/EmuKrnlXbox.cpp]: same

**Comment by PatrickvL** [comment]
Nicely done!

File: import/OpenXDK/include/xboxkrnl/xbox.h (hunk 1 above)
Comment (Sarkie) [inline import/OpenXDK/include/xboxkrnl/xbox.h]: ?

**Pull request merged**